lib/streamlit/elements/write.py

```diff
@@ -44,8 +58,183 @@
+        >>>     st.session_state.messages = []
+        >>>
+        >>> for message in st.session_state.messages:
+        >>>     st.chat_message(message["role"]).write(message["content"])
+        >>>
+        >>> if prompt := st.chat_input("What is up?"):
+        >>>     st.chat_message("user").write(prompt)
+        >>>     st.session_state.messages.append({"role": "user", "content": prompt})
+        >>>
+        >>>     with st.chat_message("assistant"):
+        >>>         response = st.experimental_stream(
+        >>>             client.chat.completions.create(
+        >>>                 model="gpt-3.5-turbo",
+        >>>                 messages=[
+        >>>                     {"role": m["role"], "content": m["content"]}
+        >>>                     for m in st.session_state.messages
+        >>>                 ],
+        >>>                 stream=True,
+        >>>             )
+        >>>         )
+        >>>        st.session_state.messages.append({"role": "assistant", "content": response})
+
+        ..  output::
+            https://doc-stream-openai.streamlit.app/
+            height: 400px
+
+        """
+
+        stream_container: DeltaGenerator | None = None
+        streamed_response: str = ""
+        written_content: List[Any] = StreamingOutput()
+
+        def flush_stream_response():
+            """Write the full response to the app."""
+            nonlocal streamed_response
+            nonlocal stream_container
+
+            if streamed_response and stream_container:
+                # Replace the stream_container element the full response
+                stream_container.write(
+                    streamed_response, unsafe_allow_html=unsafe_allow_html
+                )
+                written_content.append(streamed_response)
+                stream_container = None
+                streamed_response = ""
+
+        # Make sure we have a generator and not just a generator function.
+        stream = arg() if inspect.isgeneratorfunction(arg) else arg
+
+        # Iterate through the generator and write each chunk to the app
+        # with a type writer effect.
+        for chunk in stream:  # type: ignore
+            if type_util.is_type(
+                chunk, "openai.types.chat.chat_completion_chunk.ChatCompletionChunk"
+            ):
+                # Try to convert openai chat completion chunk to a string:
+                with contextlib.suppress(Exception):
+                    chunk = chunk.choices[0].delta.content or ""
+
+            if isinstance(chunk, str):
+                first_text = False
+                if not stream_container:
+                    stream_container = self.dg.empty()
+                    first_text = True
+                streamed_response += chunk
+                # Only add the streaming symbol on the second text chunk
+                stream_container.write(
+                    streamed_response + ("" if first_text else _TEXT_CURSOR),
```

Comment:
If I understand correctly, in case it is a `first_text` chunk, we do not show `_TEXT_CURSOR`, so for example if a sequence is `["My ", "name ", "is ", "Karen!"]` we will not show the cursor after "My ", but only starting from "name ".

Reply:
Yep, this is the intended, since it does look a bit strange to use this typewriter effect even though it is just a single text.